dagql/idproto/id.proto

```diff
@@ -11,62 +11,76 @@ option go_package = "./idproto";
 // It may be binary=>base64-encoded to be used as a GraphQL ID value for
 // objects. Alternatively it may be stored in a database and referred to via an
 // RFC-6920 ni://sha-256;... URI.
-message ID {
-  // The ID of the object that the field selection will be evaluated against.
-  //
-  // If not present, the root Query object is implied.
-  ID base = 1;
-
-  // The GraphQL type of the value.
-  Type type = 2;
-
-  // GraphQL field name.
-  string field = 3;
-
-  // GraphQL field arguments, always in alphabetical order.
-  repeated Argument args = 4;
-
-  // If true, this Selector is not reproducible.
-  //
-  // TODO: do we need to refer to session/client IDs or anything here? Or is
-  // that all internal? Forcing function is whether this is used as an
-  // in-memory query cache key. But the query cache might be made per-session
-  // or even per-client instead anyway! What buys us the most?
-  bool tainted = 5;
-
-  // If true, this Selector may be omitted from the pipeline without changing
-  // the ultimate result.
-  //
-  // This is used to prevent meta-queries like 'pipeline' and 'withFocus' from
-  // busting cache keys when desired.
-  //
-  // It is worth noting that we don't store meta information at this level and
-  // continue to force metadata to be set via GraphQL queries. It makes IDs
-  // always easy to evaluate.
-  bool meta = 6;
-
-  // If the field returns a list, this is the index of the element to select.
-  // Note that this defaults to zero, as IDs always refer to
-  //
-  // Here we're teetering dangerously close to full blown attribute path
-  // selection, but we're intentionally limiting ourselves instead to cover
-  // only the common case of returning a list of objects. The only case not
-  // handled is a nested list. Don't do that; have a type instead.
-  int64 nth = 7;
-
-  // The module that provides the implementation of the field.
-  //
-  // The actual usage of this ID is opaque to the protocol. In Dagger this is
-  // the module ID providing the implementation of the field.
-  Module module = 8;
+message RawID {
+  message Fields {
```

Comment:
Would prefer the protocol type to not be prefixed with Raw, but I can see how it helps with the rest of the code.

A bit of a context dump:

The protocol details will matter soon since the current plan is to have the frontend parse ID protobuf payloads sent over in Otel span attributes. We tried to avoid a frontend Protobuf dependency, but it's very awkward to translate arguments in particular into span attributes. We're already incentivized to have a stable ID format (#3923), and JS/TS should be good at protobuf anyway, so :man_shrugging: maybe it's fine?

Streaming the protobufs over will be 1000x better with this PR though: now we can just send infividual `Fields` over the wire, one per span, and have them all go into one big `RawID` client side, which is really :fire::sunglasses::man_cook::pinched_fingers::kiss:

In the interest of getting this protocol stable more quickly, I'm siezing an opportunity here for some bikeshedding :grin: 

While we're changing names, hopefully we can find a better one than "ID." @aluzzardi and found it too confusing while mocking up the Cloud API, and we landed on "Call" which I still like, since each of these things (currently `Fields`) corresponds to a function call in our new terminology.

So, building on Call, I dunno, "Chain?" (a...chain of function calls)

```suggestion
message Chain {
  message Call {
```

"Thunk?" (an unevaluated 0-arity function; IDs are like serializable thunks)

```suggestion
message Thunk {
  message Call {
```

Punt on it and just make "Call" the hero with a boring name for its wrapper?

```suggestion
message Envelope {
  message Call {
```

"Scope"?

```suggestion
message Scope {
  message Call {
```

I yield the floor :pray:

Reply:
Yeah `Call` SGTM.

For the "outer" type, I actually think `DAG` (or `Dag`, etc.) would be technically accurate (besides being on brand and all that 😁). It's more than a `Chain` because it stores the IDs not only for the chain of `base`s, it also has the IDs for any args of type `ID`, the `ID` of modules, etc. Each of the current `*IDDigest` fields in the types are the edges and the current `RawID_Field` type is the vertex type.

So something like:
```proto
message DAG {
  message Call {
    // The digest of the call that the field selection will be evaluated against.
    string baseCallDigest = 1;

    // ...

    // A unique digest of this Call
    string digest = 9;
  }

  // The digest of the Call for the "root" of the DAG. All other Calls
  // in this message are referenced by this one via their base, args or module.
  string rootCallDigest = 9;

  // Map of Call digests to the Call they represent. This structure
  // allows us to deduplicate occurrences of the same Call in the DAG.
  map<string, Call> callsByDigest = 10;
}

message CallModule {
  // The digest of the Call that provides the module.
  string callDigest = 1;

  // ...
}

message CallArgument {
  string name = 1;
  CallLiteral value = 2;
}

// A value passed to an argument or contained in a list.
message CallLiteral {
  oneof value {
    string callDigest = 1;
    // ...
  }
}
```

WDYT?

Reply:
> Aside from that, think I'd prefer to just keep a flat set of types without prefixing or nesting them, just to leave the door open for other shapes/configurations/reuse.

Yeah makes sense 👍

> If those prefixes are to avoid name clashes with the types used in by code, maybe we should just move the protobuf generated types into something like callpbv1?

Also SGTM, especially in the context of this being used in a wider context by the frontend, etc.

Will make these updates.

Reply:
@vito pushed that here: https://github.com/dagger/dagger/pull/6836/commits/87af4a15151bcff76246a395c01bd32b3713f0d7

* `dagql/call/callpbv1` has the proto codegen
* `dagql/call` has all the wrappers around those types

I ended up leaving the wrapper type around `callpbv1.DAG`+`callpbv1.Call` named ID. Partially because the way its used abstracts over those two types, so neither `DAG` or `Call` really fit. And partially because renaming that low-level ID type to e.g. `Call` ended making a lot of the consuming code super awkward in terms of names.

So I guess the idea is that these wrapper types are specific to the `dagql/`engine` use case and thus `ID` ends up being a clearer name. All of that is in a separate package from the pb though, so other future consumers won't be impacted.